### ISSUE TITLE
Ivy compares :action against the literal symbol 'identity

### ISSUE THREAD

**Issue by raxod502** [abo-abo/swiper#1632]
The following code emulates the operation of `find-file` with `ivy-mode` enabled. It works fine (see screenshot below):

```elisp
(ivy-read "Find file: " 'read-file-name-internal
          :predicate 'file-exists-p
          :require-match 'confirm-after-completion
          :initial-input "~/.emacs.d/init.el"
          :preselect "~/.emacs.d/init.el"
          :def "~/.emacs.d/init.el"
          :history file-name-history
          :keymap nil
          :sort t
          :dynamic-collection nil
          :caller 'read-file-name-internal
          :action 'identity)
```

![image](https://user-images.githubusercontent.com/6559064/41511560-5d2e86ec-7236-11e8-81d2-9f14307801c6.png)

The following code does the same but substituting the function `(lambda (x) x)` instead of `'identity`. This should make no difference at all, but actually it breaks Ivy:

```elisp
(ivy-read "Find file: " 'read-file-name-internal
          :predicate 'file-exists-p
          :require-match 'confirm-after-completion
          :initial-input "~/.emacs.d/init.el"
          :preselect "~/.emacs.d/init.el"
          :def "~/.emacs.d/init.el"
          :history file-name-history
          :keymap nil
          :sort t
          :dynamic-collection nil
          :caller 'read-file-name-internal
          :action (lambda (x) x))
```

![image](https://user-images.githubusercontent.com/6559064/41511564-7f02404c-7236-11e8-8405-b604f4de3253.png)

This is because there is a check against the literal symbol `'identity` in `ivy--reset-state`:

https://github.com/abo-abo/swiper/blob/6f2939485d33e9b28022d3b6912a50669dcdd596/ivy.el#L1917-L1918

**Comment by abo-abo** [comment]
> This should make no difference at all, but actually it breaks Ivy:

Your code doesn't break if you fix the missing quote:

```elisp
- :history file-name-history
+ :history 'file-name-history
```

> This is because there is a check against the literal symbol 'identity in ivy--reset-state:

Checking for `'identity` is this is part of the API: the user passes `'identity` if they only care about the return result of `ivy-read`.
Now for `find-alternate-file`, nil is passed for `:action` and this should be handled fine as well. Passing `(lambda (x) x)` isn't the way to have `ivy-read` not call the action - the action will still be called as expected.

**Comment by raxod502** [comment]
> Your code doesn't break if you fix the missing quote:

Actually, it does still break. Exactly the same thing happens with or without that quote.

> Checking for `'identity` is this is part of the API

It is? Where is that documented?

> the user passes `'identity` if they only care about the return result of `ivy-read`.

OK, fine, but why would that affect the operation of `:initial-input`?

> Now for find-alternate-file, nil is passed for `:action`

Not in my setup. I use `ivy-prescient.el`, which adds an advice to `ivy-read` causing the `:action` to always be non-nil.

> Passing `(lambda (x) x)` isn't the way to have `ivy-read` not call the action - the action will still be called as expected.

Yes, of course. But I must be misunderstanding something here. Firstly, if the action is `(lambda (x) x)`, then why would it make a difference whether the action is called or not? It can't possibly _do_ anything. And secondly, how could whether or not the action is called affect the operation of `:initial-input`?

**Comment by raxod502** [comment]
FWIW, the actual problem in my setup is that `find-alternate-file` is broken. I only factored out the above example for reproducibility.

**Comment by abo-abo** [comment]
> Actually, it does still break. Exactly the same thing happens with or without that quote.

Can you give me a reproducible scenario, since I can't reproduce with `make plain`?

**Comment by raxod502** [comment]
I'm using revision 80f05a7845085a24938d1db1753b53b282109281 of the `swiper` repository. I ran `make plain`, pasted the second example from my original post with the missing quote added, i.e.

```elisp
(ivy-read "Find file: " 'read-file-name-internal
          :predicate 'file-exists-p
          :require-match 'confirm-after-completion
          :initial-input "~/.emacs.d/init.el"
          :preselect "~/.emacs.d/init.el"
          :def "~/.emacs.d/init.el"
          :history 'file-name-history
          :keymap nil
          :sort t
          :dynamic-collection nil
          :caller 'read-file-name-internal
          :action (lambda (x) x))
```

pressed `C-x C-e`, and got exactly the result I showed above.

Do you not have a file called `~/.emacs.d/init.el` on your machine? That is the only way I can think that our environments could be different in a relevant way, except perhaps Emacs version (I'm running 26.1).

**Comment by abo-abo** [comment]
> Do you not have a file called ~/.emacs.d/init.el on your machine?

No.

> except perhaps Emacs version (I'm running 26.1).

Using 26.1 on Ubuntu 16. No error is reproduced. Instead, I'm getting the input "~/.emacs.d/init.el" in the minibuffer and `~/.emacs.d/init.el` as the only candidate. Selecting it makes <kbd>C-x C-e</kbd> return the expanded file name.

**Comment by basil-conto** [comment]
FWIW, you can just use `user-init-file` instead of a hard-coded string.

**Comment by raxod502** [comment]
Even if you create `~/.emacs.d/init.el` (or replace the path with one that does exist), you still don't get the behavior I described?

I get the same result with `~/.emacs.d/<some-file-that-doesn't-exist>` and `~/<some-folder-that-doesn't-exist>/<some-file>`.

Also, I tried the same thing (with `~/.emacs.d/init.el` existing) on Arch Linux. Same result.

Also, I found another machine running a newer version of macOS, did a fresh `brew install emacs --with-cocoa`, cloned just the Swiper repository, and tested again. Same result. There's no `~/.emacs.d/init.el` on that machine.

Frankly I've yet to find a situation in which the problem _isn't_ reproduced.
